File: src/test/java/de/schramm/royalbash/controller/service/core/SpotTest.java
package de.schramm.royalbash.controller.service.core;

import de.schramm.royalbash.controller.service.core.creature.NoOpCreature;
import lombok.val;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class SpotTest {

    @Test
    public void should_deliver_placed_creature() {

        // Given
        val creature = NoOpCreature.builder().build();
        val testee = Spot.builder().build();

        // When
        val spot = testee.place(creature);

        // Then
        assertThat(spot.getCreature().isPresent()).isTrue();
        assertThat(spot.getCreature().get()).isEqualTo(creature);
    }

    @Test
    public void should_update_creature() {

        // Given
        val creature = NoOpCreature.builder().build();
        val otherCreature = creature.toBuilder()
                .hitpoints(15)
                .build();
        val testee = Spot.builder()
                .creature(creature)
                .build();

        // When
        val spot = testee.updateCreature(creature, otherCreature);

        // Then
        assertThat(spot.getCreature().isPresent()).isTrue();
        assertThat(spot.getCreature().get()).isEqualTo(otherCreature);
    }

    @Test
    public void should_not_update_creature() {

        // Given
        val creature = NoOpCreature.builder().build();
        val testee = Spot.builder().build();

        // When
        val spot = testee.updateCreature(creature, creature);

        // Then
        assertThat(spot.getCreature().isPresent()).isFalse();
    }

    @Test
    public void should_remove_creature() {

        // Given
        val creature = NoOpCreature.builder().build();
        val testee = Spot.builder()
                .creature(creature)
                .build();

        // When
        val spot = testee.removeCreature(creature);

        // Then
        assertThat(spot.getCreature().isPresent()).isFalse();
    }

    @Test
    public void should_not_remove_creature() {

        // Given
        val creature = NoOpCreature.builder().build();
        val otherCreature = NoOpCreature.builder()
                .hitpoints(15)
                .build();
        val testee = Spot.builder()
                .creature(creature)
                .build();

        // When
        val spot = testee.removeCreature(otherCreature);

        // Then
        assertThat(spot.getCreature().isPresent()).isTrue();
        assertThat(spot.getCreature().get()).isEqualTo(creature);
    }
}

File: src/test/java/de/schramm/royalbash/controller/service/core/SpotTest.kt
package de.schramm.royalbash.controller.service.core

import de.schramm.royalbash.controller.service.core.card.creature.CreatureMock
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.*

class SpotTest {

    @Test
    fun should_deliver_placed_creature() {

        // Given
        val creature = CreatureMock("Id 1")
        val testee = Spot.builder().build()

        // When
        val spot = testee.place(creature)

        // Then
        assertThat(spot.creature).isEqualTo(Optional.of(creature))
    }

    @Test
    fun should_update_creature() {

        // Given
        val creature = CreatureMock("Id 1")
        val otherCreature = creature.copy(hitpoints = 15)
        val testee = Spot.builder()
                .creature(creature)
                .build()

        // When
        val spot = testee.updateCreature(creature, otherCreature)

        // Then
        assertThat(spot.creature).isEqualTo(Optional.of(otherCreature))
    }

    @Test
    fun should_not_update_creature() {

        // Given
        val creature = CreatureMock("Id 1")
        val testee = Spot.builder().build()

        // When
        val spot = testee.updateCreature(creature, creature)

        // Then
        assertThat(spot.creature).isNotPresent
    }

    @Test
    fun should_remove_creature() {

        // Given
        val creature = CreatureMock("Id 1")
        val testee = Spot.builder()
                .creature(creature)
                .build()

        // When
        val spot = testee.removeCreature(creature)

        // Then
        assertThat(spot.creature).isNotPresent
    }

    @Test
    fun should_not_remove_creature() {

        // Given
        val creature = CreatureMock("Id 1")
        val otherCreature = CreatureMock("Id 2", hitpoints = 15)
        val testee = Spot.builder()
                .creature(creature)
                .build()

        // When
        val spot = testee.removeCreature(otherCreature)

        // Then
        assertThat(spot.creature).isEqualTo(Optional.of(creature))
    }
}